### DISCUSSION TITLE
guides/object-cropping/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#12667] [Docs]
# guides/object-cropping/

Learn how to isolate and extract specific objects from images and videos using YOLOv8 object cropping.

### Code

```python
import os

import cv2

from ultralytics import YOLO
from ultralytics.utils.plotting import Annotator, colors

model = YOLO("yolo11n.pt")
names = model.names

cap = cv2.VideoCapture("path/to/video/file.mp4")
assert cap.isOpened(), "Error reading video file"
w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

crop_dir_name = "ultralytics_crop"
if not os.path.exists(crop_dir_name):
    os.mkdir(crop_dir_name)

# Video writer
video_writer = cv2.VideoWriter("object_cropping_output.avi", cv2.VideoWriter_fourcc(*"mp4v"), fps, (w, h))

idx = 0
while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        print("Video frame is empty or video processing has been successfully completed.")
        break

    results = model.predict(im0, show=False)
    boxes = results[0].boxes.xyxy.cpu().tolist()
    clss = results[0].boxes.cls.cpu().tolist()
    annotator = Annotator(im0, line_width=2, example=names)

    if boxes is not None:
        for box, cls in zip(boxes, clss):
            idx += 1
            annotator.box_label(box, color=colors(int(cls), True), label=names[int(cls)])

            crop_obj = im0[int(box[1]) : int(box[3]), int(box[0]) : int(box[2])]

            cv2.imwrite(os.path.join(crop_dir_name, str(idx) + ".png"), crop_obj)

    cv2.imshow("ultralytics", im0)
    video_writer.write(im0)

    if cv2.waitKey(1) & 0xFF == ord("q"):
        break

cap.release()
video_writer.release()
cv2.destroyAllWindows()
```

### Docs

https://docs.ultralytics.com/guides/object-cropping/

**Comment by HsnKmlMt** [comment]
Hi! How can i specify classed to be cropped like for my case i just want people.

**Reply by glenn-jocher** [reply]
Hello! To specify that you only want to crop out people from your video using YOLOv8, you'll need to filter the detections by the class corresponding to people. Typically, in the COCO dataset which YOLO models often train on, 'person' is class 0.

Here's a modified snippet from the example you may find helpful:

```python
from ultralytics import YOLO
import cv2
import os

model = YOLO("yolov8n.pt")
person_class_id = model.names.index('person')  # Ensure 'person' is in the model's class names

cap = cv2.VideoCapture("path/to/video/file.mp4")
assert cap.isOpened(), "Error reading video file"

crop_dir_name = "ultralytics_crop"
if not os.path.exists(crop_dir_name):
    os.mkdir(crop_dir_name)

idx = 0
while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        break

    results = model(im0)
    boxes = results.boxes.xyxy[results.boxes.cls == person_class_id].cpu().tolist()  # Filter for person class

    for box in boxes:
        idx += 1
        crop_obj = im0[int(box[1]):int(box[3]), int(box[0]):int(box[2])]
        cv2.imwrite(os.path.join(crop_dir_name, f"person_{idx}.png"), crop_obj)

cap.release()
cv2.destroyAllWindows()
```

This code will crop and save images only for detected persons, assuming 'person' is the correct class label and indexed correctly in your model. Adjust according to your specific classes and needs! 👍

**Reply by HsnKmlMt** [reply]
Thank you, I really appreciate it!

**Reply by glenn-jocher** [reply]
Hello! We're really glad you found the materials helpful! If you ever need to specifically crop specific classes, such as 'people', you can do so by filtering the detections based on class IDs before cropping. Here's a quick example on how you might modify the example to only crop 'people':

```python
from ultralytics import YOLO
import cv2
import os

model = YOLO("yolov8n.pt")  # Load your model
person_class_id = model.names.index('person')  # Assuming 'person' is the class name for people

cap = cv2.VideoCapture("path/to/video.mp4")
assert cap.isOpened(), "Error reading video file"

while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break
    results = model.predict(frame)
    boxes = results.boxes.xyxy[results.boxes.cls == person_class_id].cpu().tolist()  # Filter detections to only 'person'

    for idx, box in enumerate(boxes):
        x1, y1, x2, y2 = map(int, box[:4])
        crop_img = frame[y1:y2, x1:x2]
        cv2.imwrite(f'cropped_person_{idx}.jpg', crop_img)
```

Feel free to replace 'person' with any class you're interested in, and make sure the class name matches what's defined in your model's class names. Let us know if you need further assistance! 🚀

**Reply by umairrkhn** [reply]
Hey Glenn,
I am getting this error using your implementation:

person_class_id = model.names.index('person')  # Ensure 'person' is in the model's class names
                      ^^^^^^^^^^^^^^^^^
AttributeError: 'dict' object has no attribute 'index'

**Reply by glenn-jocher** [reply]
Hey there!

It looks like you're trying to use the `.index()` method on a dictionary, which isn't supported. The `model.names` in YOLOv8 is a list, not a dictionary. Make sure that `model.names` is correctly referenced and is a list. Here’s how you can do it:

```python
model = YOLO("yolov8n.pt")
person_class_id = model.names.index('person')  # This should work if model.names is a list
```

If `model.names` is somehow being altered to a dictionary in your code, you'll need to ensure it remains a list or adjust your approach accordingly. If the issue persists, could you share more of your code context? That might help diagnose the issue better. 😊

**Reply by glenn-jocher** [reply]
Hey there! It looks like you're trying to use the `index()` method on a dictionary, which isn't supported. The `model.names` is likely a dictionary, not a list. You can access the class index directly if you know the class name. Here’s how you can modify your code:

```python
person_class_id = model.names['person']  # Directly access the 'person' class index
```

Make sure that 'person' is indeed the key for the class you're interested in. If `model.names` is structured differently, you might need to adjust the key accordingly. Let me know if this helps or if you need further assistance! 😊

**Reply by zakirh** [reply]
To those who are facing issue on finding person class:

person_class_id = None
for class_id, class_name in model.names.items():
    if class_name =='person':
        person_class_id = class_id
        break

if person_class_id is None:
    print("The 'person' class was not found.")
else:
    print(f"'person' class ID: {person_class_id}")

**Reply by glenn-jocher** [reply]
@zakirh thank you for sharing this helpful approach to finding the class ID for 'person.' For others who might encounter similar issues, make sure that `model.names` is checked as either a list or dictionary depending on the YOLO model version being used. Your solution will work seamlessly when `model.names` is a dictionary. If any further clarification is needed, feel free to ask!

**Comment by AdityaJaiswal99** [comment]
How can we change the frame height and width to a specific pixel and change FPS?

**Reply by pderrenger** [reply]
@AdityaJaiswal99 hi there!

To change the frame height, width, and FPS of a video using OpenCV, you can set these properties directly on the `cv2.VideoCapture` and `cv2.VideoWriter` objects. Below is an example of how you can achieve this:

```python
import cv2

# Load the video
cap = cv2.VideoCapture("path/to/video/file.mp4")
assert cap.isOpened(), "Error reading video file"

# Set the desired frame width, height, and FPS
frame_width = 640  # Set your desired width
frame_height = 480  # Set your desired height
fps = 30  # Set your desired FPS

# Create a VideoWriter object to save the modified video
video_writer = cv2.VideoWriter("modified_video.avi", cv2.VideoWriter_fourcc(*"mp4v"), fps, (frame_width, frame_height))

while cap.isOpened():
    success, frame = cap.read()
    if not success:
        break

    # Resize the frame to the desired dimensions
    resized_frame = cv2.resize(frame, (frame_width, frame_height))

    # Write the resized frame to the output video
    video_writer.write(resized_frame)

    # Display the frame (optional)
    cv2.imshow("Resized Video", resized_frame)
    if cv2.waitKey(1) & 0xFF == ord("q"):
        break

# Release resources
cap.release()
video_writer.release()
cv2.destroyAllWindows()
```

This script reads a video, resizes each frame to the specified width and height, and writes the modified frames to a new video file with the desired FPS.

For more detailed information on object cropping using YOLOv8, you can refer to our [Object Cropping Guide](https://docs.ultralytics.com/guides/object-cropping/).

Feel free to reach out if you have any more questions! 😊

**Comment by AdityaJaiswal99** [comment]
Can we make the model, so it does not take the same pictures in the frame again?

**Reply by glenn-jocher** [reply]
@AdityaJaiswal99 thank you for your question! If you're looking to avoid capturing the same objects multiple times within a frame, you can leverage the Non-Maximum Suppression (NMS) feature of YOLOv8. NMS helps in eliminating redundant bounding boxes by keeping only the highest confidence detections and discarding others that overlap significantly.

Here's a brief example of how you can adjust the `iou` threshold to refine your detections and reduce duplicates:

```python
from ultralytics import YOLO

# Load the model
model = YOLO("yolov8n.pt")

# Set the IoU threshold to a lower value to reduce overlapping detections
results = model.predict("path/to/your/image_or_video", iou=0.5)

# Process the results as needed
for result in results:
    boxes = result.boxes.xyxy.cpu().tolist()
    clss = result.boxes.cls.cpu().tolist()
    # Further processing...
```

In this example, setting the `iou` parameter to a lower value (e.g., 0.5) helps in reducing the number of overlapping boxes, thus minimizing the chances of capturing the same object multiple times.

For a more comprehensive guide on object cropping and additional tips, please visit our [Object Cropping Guide](https://docs.ultralytics.com/guides/object-cropping/).

If you encounter any issues or have further questions, feel free to provide a reproducible example, and ensure you are using the latest version of the packages. This will help us assist you more effectively. 😊

Happy coding!

**Comment by NOSPAA21** [comment]
Can i use it so as to train a model especially for printed circuit board defect detection to find defects in certain regions of a pcb?Would this be helpful?

**Reply by RizwanMunawar** [reply]
@NOSPAA21 Yes, you can train the Ultralytics YOLO11 model on the circuit board defect detection dataset by following the [training](https://docs.ultralytics.com/modes/train/) documentation. Once the model is trained, you can use the [object-cropping](https://docs.ultralytics.com/guides/object-cropping/) solution to crop the different defects that the model detects.
